source/common/router/router.cc

```diff
@@ -760,31 +862,42 @@ void Filter::onUpstreamHeaders(uint64_t response_code, Http::HeaderMapPtr&& head
     upstream_request.upstream_host_->healthChecker().setUnhealthy();
   }
 
+  bool could_not_retry = false;
+
+  // Check if this upstream request was already retried, for instance after
+  // hitting a per try timeout. Don't retry it if we already have.
   if (retry_state_) {
-    // Notify retry modifiers about the attempted host.
-    retry_state_->onHostAttempted(upstream_request.upstream_host_);
-
-    // Capture upstream_host since setupRetry() in the following line will clear
-    // upstream_request.
-    const auto upstream_host = upstream_request.upstream_host_;
-    const RetryStatus retry_status =
-        retry_state_->shouldRetryHeaders(*headers, [this]() -> void { doRetry(); });
-    if (retry_status == RetryStatus::Yes && setupRetry(end_stream)) {
-      Http::CodeStats& code_stats = httpContext().codeStats();
-      code_stats.chargeBasicResponseStat(cluster_->statsScope(), "retry.",
-                                         static_cast<Http::Code>(response_code));
-      upstream_host->stats().rq_error_.inc();
-      return;
-    } else if (retry_status == RetryStatus::NoOverflow) {
-      callbacks_->streamInfo().setResponseFlag(StreamInfo::ResponseFlag::UpstreamOverflow);
-    } else if (retry_status == RetryStatus::NoRetryLimitExceeded) {
-      callbacks_->streamInfo().setResponseFlag(
-          StreamInfo::ResponseFlag::UpstreamRetryLimitExceeded);
+    if (upstream_request.retried_) {
+      // We already retried this request (presumably for a per try timeout) so
+      // we definitely won't retry it again. Check if we would have retried it
+      // if we could.
+      could_not_retry = retry_state_->wouldRetryFromHeaders(*headers);
+    } else {
+      const RetryStatus retry_status =
+          retry_state_->shouldRetryHeaders(*headers, [this]() -> void { doRetry(); });
+      // Capture upstream_host since setupRetry() in the following line will clear
+      // upstream_request.
+      const auto upstream_host = upstream_request.upstream_host_;
+      if (retry_status == RetryStatus::Yes && setupRetry()) {
+        if (!end_stream) {
+          upstream_request.resetStream();
+        }
+        upstream_request.removeFromList(upstream_requests_);
+
+        Http::CodeStats& code_stats = httpContext().codeStats();
+        code_stats.chargeBasicResponseStat(cluster_->statsScope(), "retry.",
+                                           static_cast<Http::Code>(response_code));
+        upstream_host->stats().rq_error_.inc();
+        return;
+      } else if (retry_status == RetryStatus::NoOverflow) {
```

Comment:
it's enough different that' it'd be difficult. We call a different retry state function, reset the stream on headers but not soft per try timeout, charge http code on headers and not soft per try timeout, and also have the `could_not_retry` logic which tells us whether we're done (e.g. this was the last in-flight attempt) vs need to wait for other attempts to potentially get headers back.

I'll take a look into deduplicating the retry handing `onSoftPerTryTimeout` with `maybeRetryReset` though, those look similar enough.

Reply:
after looking at it, it's probably not worth it because you end up with a nested lambda situation (one calls `shouldRetryReset`, one calls `shouldRetryHedgedPerTryTimeout` and both need to provide the `doRetry()` lambda)